src/rust/engine/src/externs/options.rs

```diff
@@ -177,19 +175,50 @@ type OptionValueDerivation<T> = Vec<(T, Vec<isize>)>;
 // We could get rid of this tuple type by representing the final value and its rank as
 // a singleton derivation (in the case where we don't otherwise need the full derivation).
 // But that would allocate two unnecessary Vecs for every option.
-type OptionValue<T> = (Option<T>, isize, Option<OptionValueDerivation<T>>);
+type OptionValue<'py, T> = (Option<T>, isize, Option<OptionValueDerivation<'py, T>>);
+
+fn source_to_details(source: &Source) -> Option<&str> {
+    match source {
+        Source::Default => None,
+        Source::Config { ordinal: _, path } => Some(path),
```

Comment:
I dispensed with the prefix "from" in these, since it adds no information, and in the case of config file paths it would require storing a "from <path>" string somewhere with an appropriate lifetime, and there is no obvious place for that (other than on the Source itself, but that seems silly).